[16.0][MIG] sale_order_import: Migration to 16.0

QuocDuong1306 opened the pull request.

alexis-via committed:
Initial check-in of module sale_order_import, sale_stock_order_import, sale_order_import_ubl, sale_order_import_csv, sale_order_ubl, sale_commercial_partner

alexis-via committed:
Add 2 README.rst

Add another README.

alexis-via committed:
FIX Don't use _inherit = ['business.document.import']

because we want to have access to the code of the modules that inherit business.document.import

Commit:
[IMP] handle the case where the xml file is generated with mime type
'text/xml'

alexis-via committed:
Add support for extraction + matching on website

PEP8 fix

Commit:
8.0 Add support for partner bank matching on invoice update (#6)

Add support for partner bank matching on invoice update (before, it was only supported on invoice creation)

oca-transbot committed:
OCA Transbot updated translations from Transifex

alexis-via committed:
Prepare v10 branch

Rename __openerp__.py to __manifest__.py and set installable to False

alexis-via committed:
Continue port of modules for v10.0, in particular sale_order_import_* module

Fix spelling mistake and other remarks on README by Tarteo

alexis-via committed:
Port account_invoice_import_factur-x and account_invoice_import_ubl to v10.0

base_business_document_import: Add support for the creation of res.bank

alexis-via committed:
Port purchase_order_import* to v10.0

Add ubl invoice generation option in accounting config page

alexis-via committed:
FIX crash when pivot format had a 'note' key (and no attachment)

Add method to create SO in sale.order.import accessible via JSON-RPC

oca-transbot committed:
OCA Transbot updated translations from Transifex

tarteo committed:
[11.0][MIG] sale_order_import: Migration + add tests

Add readme fragments

alexis-via committed:
Fix import of sale order with price_source='pricelist' when pricelist has visible discounts

Code improvements in sale_order_import
Add unit tests in sale_order_import
Use display_name instead of name_get()[0][1]

Fix travis

alexis-via committed:
Check the VAT number of the destination partner, to make sure the business document is imported in the right company

oca-travis committed:
[UPD] Update sale_order_import.pot

OCA-git-bot committed:
[ADD] icon.png

astirpe committed:
[12.0][MIG] sale_order_import

sebalix committed:
[IMP] sale_order_import: black, isort, prettier

sebalix committed:
[MIG] sale_order_import: Migration to 13.0

sebalix committed:
[UPD] README.rst

oca-travis committed:
[UPD] Update sale_order_import.pot

OCA-git-bot committed:
[UPD] README.rst

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: edi-13.0/edi-13.0-sale_order_import
Translate-URL: https://translation.odoo-community.org/projects/edi-13-0/edi-13-0-sale_order_import/

jbaudoux committed:
Parse and import Sales Order Invoicing Partner. Refactor shipping partner matching.

A shipping partner is not especially related to the ordering party. It
could be any partner. However, we want to ensure that the address is
respected.

Fix-up: PartyIdentification/ID is a 1..N relation. Parse each of them for a match

jbaudoux committed:
sale_order_import: Match existing quotation respecting the PO number

Do not suggest to update an open quotation that does not have the
corresponding client ref. UBL endpoint was blocked as soon as a
quotation for the same customer was existing.

jbaudoux committed:
sale_order_import: prevent to create 2 SO with the same ref and commercial entity

OCA-git-bot committed:
sale_order_import 13.0.2.0.0

phuctranfxvn committed:
[IMP] sale_order_import: black, isort, prettier

TDu committed:
[13.0][IMP] sale_order_import add hook search existing

When importing an order the module checks if a similar order already
exists in the system. The search domain used for this is not
customizable.
This change extract the domain computation for the search in its own
function.

TDu committed:
[FW][14.0][IMP] Add sale order ubl parse latest delivery

Parse the latest delivery date and time from the delivery node of an UBL
sales order.
And assign the value to the `commitment_date` of the sales order.

phuctranfxvn committed:
[MIG] sale_order_import: Migration to 14.0

TDu committed:
[ADD][13.0]sale_order_import no mail subscribe

Disable the subscription (mail chatter) to the new sales order
for the user used by the endpoint.
To me it does make sense because good practice would be to have a
specific user for the endpoint.

oca-travis committed:
[UPD] Update sale_order_import.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
sale_order_import 14.0.1.1.0

simahawk committed:
sale_order_import: split xml parsing and ease error handling

simahawk committed:
sale_order_import: parser_xml_order fix not implemented exception

simahawk committed:
sale_order_import: fix action lookup

oca-travis committed:
[UPD] Update sale_order_import.pot

OCA-git-bot committed:
sale_order_import 14.0.1.1.1

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: edi-14.0/edi-14.0-sale_order_import
Translate-URL: https://translation.odoo-community.org/projects/edi-14-0/edi-14-0-sale_order_import/

OCA-git-bot committed:
sale_order_import 14.0.1.2.0

simahawk committed:
sale_order_import: fix deprecation warning

simahawk committed:
sale_order_import: fix XMLSyntaxError handling

OCA-git-bot committed:
sale_order_import 14.0.1.2.1

simahawk committed:
sale_order_import: _parse_xml handle NotImplementedError

When no handler is defined for the given document
NotImplementedError could be raised.

Handle this case properly when parsing.

OCA-git-bot committed:
sale_order_import 14.0.1.2.2

TDu committed:
sale_order_import: add line update values hook

simahawk committed:
sale_order_import: fix missing company default

SO and SOL creation should be done w/ all defaults loaded.
In particular, the company is not propagated directly to the SOL
when we play onchanges.

Hence, at least in our tests, it can happen
that the company is not set yet on the line
and the computation of the price via 'product._get_tax_included_unit_price'
is broken when calling 'res.currency._convert' w/ an empty company.

Error example::

    [...]/base/models/res_currency.py", line 192, in _convert
    assert company, "convert amount from unknown company"
    AssertionError: convert amount from unknown company

ATM I'm adding only the company since it turns out that
if we load all defaults w/ `order.default_get` it breaks price
computation.

simahawk committed:
sale_order_import: reduce _prepare_order complexity

simahawk committed:
sale_order_import: refactor tests setup

simahawk committed:
sale_order_import: test onchange validation

simahawk committed:
sale_order_import: order_file_change improve readability

simahawk committed:
sale_order_import: get_xml_files_from_pdf does not exists

simahawk committed:
sale_order_import: test _parse_xml

OCA-git-bot committed:
sale_order_import 14.0.1.2.3

simahawk committed:
sale_order_import: update autorship

simahawk committed:
sale_order_import: use pdf_helper

simahawk committed:
sale_order_import: add tests TODOs

simahawk committed:
sale_order_import: test parse_pdf_order

simahawk committed:
sale_order_import: re-organize test files

simahawk committed:
sale_order_import: refactor parse_order, de-dup code

OCA-git-bot committed:
sale_order_import 14.0.1.3.0

Commit:
[UPD] Update sale_order_import.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
sale_order_import 14.0.1.4.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: edi-14.0/edi-14.0-sale_order_import
Translate-URL: https://translation.odoo-community.org/projects/edi-14-0/edi-14-0-sale_order_import/

TDu committed:
sale_order_import: fix invoicing partner

OCA-git-bot committed:
sale_order_import 14.0.1.4.1

simahawk committed:
sale_order_import: propagate all available fields

Handle additional fields dynamically if available.
This way, if you add a field to a record and it's value is injected by a parser
you won't have to override  to let it propagate.

OCA-git-bot committed:
sale_order_import 14.0.1.4.2

simahawk committed:
sale_order_import: accept default order data via ctx

You can now pass 'sale_order_import__default_vals' key in the ctx
to pass order values down the stack.

OCA-git-bot committed:
sale_order_import 14.0.1.5.0

OCA-git-bot committed:
[UPD] README.rst

Commit:
[IMP] sale_order_import: pre-commit stuff

simahawk commented on sale_order_import/wizard/sale_order_import.py:
```diff
@@ -0,0 +1,663 @@
+        [("import", "Import"), ("update", "Update")], default="import"
+    )
+    partner_id = fields.Many2one("res.partner", string="Customer")
+    csv_import = fields.Boolean(default=False, readonly=True)
```
This flag should be removed and a new selection field `import_type` should be added.
`supported_types` below should depend on this and a `_parse_{import_type}` should be used by `_parse_file`

simahawk commented on sale_order_import/wizard/sale_order_import.py:
```diff
@@ -0,0 +1,663 @@
+            error_msg = _("Unsupported XML document")
+        return xml_root, error_msg
+
+    # FIXME: not used at all
```
FIXME pls :)

simahawk commented on sale_order_import/wizard/sale_order_import.py:
```diff
@@ -0,0 +1,663 @@
+            )
+        )
+
+    # FIXME: not used at all
```
FIXME pls

simahawk commented on sale_order_import/wizard/sale_order_import.py:
```diff
@@ -0,0 +1,663 @@
+            )
+        )
+
+    # TODO: move it out to a PDF support module
```
this is a new import type  in theory. Yet, as is still based on xml parsing in the end, we can leave it here but refactored according to the "import_type" mentioned above

simahawk commented on sale_order_import/wizard/sale_order_import.py:
```diff
@@ -0,0 +1,663 @@
+        )
+        return action
+
+    # TODO: add tests
```
TODO

simahawk commented on sale_order_import/models/sale.py:
```diff
@@ -0,0 +1,28 @@
+            for (sale_id, name) in res:
+                sale = self.browse(sale_id)
+                # I didn't find a python method to easily display
+                # a float + currency symbol (before or after)
```
the ir.qweb.field.monetary widget should be used probably

simahawk replied:
can you add at least a TODO for my comment above :arrow_up:

simahawk reviewed (changes requested):
if you add the changes that I proposed, pls make them in separated commits :pray:

simahawk commented:
/ocabot migration sale_order_import

Commit:
[MIG] sale_order_import: Migration to 16.0

Commit:
[IMP] sale_order_import: Don't support importing csv anymore

Commit:
[IMP] sale_order_import: Refactor code

- Add new selection import_type field for choosing the type to import
- Refactor _parse_file with import_type field
- Drop some not used method

simahawk commented on sale_order_import/tests/test_parse_validate.py:
```diff
@@ -15,18 +15,17 @@
+            b"<?xml version='1.0' encoding='utf-8'?><root><foo>baz</foo></root>"
+        )
+        order_filename = "test_order.xml"
+        with self.assertRaises(exceptions.UserError):
```
pls check for the specific error msg if possible

simahawk replied:
?

simahawk commented on sale_order_import/wizard/sale_order_import.py:
```diff
@@ -72,18 +77,43 @@ def _parse_file(self, filename, filecontent, detect_doc_type=False):
         logger.debug("Order file mimetype: %s", filetype)
         mimetype = filetype[0]
         supported_types = {
```
to ease plugging new import types, this should stay in a separated method :wink:

simahawk commented on sale_order_import/wizard/sale_order_import.py:
```diff
@@ -72,18 +77,43 @@ def _parse_file(self, filename, filecontent, detect_doc_type=False):
+                )
+            )
+        # Call the appropriate parsing method based on the selected import type
+        if self.import_type == "xml":
```
IMO these checks are not needed. If yes, do them in their specific parsers

simahawk commented on sale_order_import/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,7 @@
+* module *sale_order_import_ubl*: adds support for `Universal Business Language (UBL) <http://ubl.xml.org/>`_ RFQs and orders as:
+
+  - XML file,
+  - PDF file with an embedded XML file.
```
mention that you can also plug additional formats by extending the wizard :wink:

simahawk reviewed (changes requested):
Thanks for the improvements! Few remarks...

Commit:
[IMP] sale_order_import: Add tests for create/update order button

simahawk reviewed (approved):
LG overall

QuocDuong1306 commented:
@simahawk , I updated

simahawk reviewed (approved):
LG

yvaucher reviewed (approved):
LGTM

yvaucher commented:
/ocabot merge nobump

yvaucher commented:
Note: we can search for a better method for the TODO about currency later, I'm not aware of one neither.

yvaucher commented:
/ocabot merge nobump
